ocaml/lang/security/marshal.yaml

```diff
@@ -0,0 +1,21 @@
+rules:
+  - id: ocamllint-marshal
+    pattern-either:
+      - pattern: input_value
+      - pattern: Marshal.from_channel
+      - pattern: Marshal.from_bytes
+      - pattern: Marshal.from_string
+    # https://eternal.red/2021/secure-ocaml-sandbox/
+    message: Marshaling is currently not type-safe and can lead to insecure behaviour when untrusted data is marshalled. Marshalling can lead to out-of-bound reads as well.
+    languages: [ocaml]
+    severity: WARNING
+    metadata:
+      category: security
+      technology:
+        - ocaml
+      cwe: "CWE-502: Deserialization of Untrusted Data"
+      confidence: HIGH
+      likelihood: MEDIUM
+      impact: HIGH
+      subcategory:
+        - vuln
```

Comment:
I don't think in its current state, this is a vuln finding rule with confidence high. For that we would need to use taint mode and add a list of sources of potential user input. Currently this rule only verifies whether data marshalled. Any checks for whether that data is 'untrusted' are missing.

Alternatively, you could add a pattern-not to exclude hardcoded  strings, and change the confidence to medium or low.

Reply:
I tried to address your inputs where I was able to. Not sure about how to exclude the hardcoded strings properly. Do you have any pointers on that? I am travelling during the next two weeks and might not be able to respond.

Reply:
Yes ofcourse! I am not too familiar with OCaml, and it seems like your testcode is missing a few tests for the different patterns you have defined. But I imagine a pattern like this will do the trick:

```
patterns:
  - pattern: Marshal.from_string $STR $INT
  - pattern-not: Marshal.from_string "..." $INT
```

This will allow you to look for calls to the from_string and exclude the ones where the string is hardcoded.

Reply:
I think for this one the pattern-not does not really make sense, since deserializing static strings is not really a useful thing to do.